File: ASMR.Web/ClientApp/src/services/GateService.ts
import { serialize } from "object-to-formdata"
import RegistrationRequestModel from "asmr/core/request/RegistrationRequestModel"
import ResendEmailAddressConfirmationRequestModel from "asmr/core/request/ResendEmailAddressConfirmationRequestModel"
import ConfirmEmailAddressRequestModel from "asmr/core/request/ConfirmEmailAddressRequestModel"
import SignInRequestModel from "asmr/core/request/SignInRequestModel"
import ForgetPasswordRequestModel from "asmr/core/request/ForgetPasswordRequestModel"
import ResetPasswordRequestModel from "asmr/core/request/ResetPasswordRequestModel"
import AuthenticationResponseModel from "asmr/core/response/AuthenticationResponseModel"
import ServiceBase from "asmr/services/ServiceBase"
import ResponseModelBase from "asmr/core/common/ResponseModelBase"

class GateService extends ServiceBase {
	private readonly path = "/api/Gate"

	public register(body: RegistrationRequestModel, imageFile: File | null, captchaResponseToken: string | null) {
		const formData = serialize(body)
		if (imageFile) {
			formData.append("image", imageFile, imageFile.name)
		}
		return this.httpClient
			.post<ResponseModelBase>(`${this.path}/register`, formData, {
				params: { captchaResponseToken },
			})
			.then(this.processResponse.bind(this))
			.finally(this.finalize.bind(this))
	}

	public resendEmailAddressConfirmation(body: ResendEmailAddressConfirmationRequestModel) {
		return this.httpClient
			.post<ResponseModelBase>(`${this.path}/email-address/resend-confirmation`, body)
			.then(this.processResponse.bind(this))
			.finally(this.finalize.bind(this))
	}

	public confirmEmailAddress(body: ConfirmEmailAddressRequestModel) {
		return this.httpClient
			.post<ResponseModelBase>(`${this.path}/email-address/confirm`, body)
			.then(this.processResponse.bind(this))
			.finally(this.finalize.bind(this))
	}

	public authenticate(body: SignInRequestModel) {
		return this.httpClient
			.post<AuthenticationResponseModel>(`${this.path}/authenticate`, body)
			.then(this.processResponse.bind(this))
			.finally(this.finalize.bind(this))
	}

	public clearSession() {
		return this.httpClient
			.post<AuthenticationResponseModel>(`${this.path}/exit`)
			.then(this.processResponse.bind(this))
			.finally(this.finalize.bind(this))
	}

	public forgetPassword(body: ForgetPasswordRequestModel, captchaResponseToken: string | null) {
		return this.httpClient
			.post<ResponseModelBase>(`${this.path}/password/forget`, body, {
				params: { captchaResponseToken },
			})
			.then(this.processResponse.bind(this))
			.finally(this.finalize.bind(this))
	}

	public resetPassword(body: ResetPasswordRequestModel) {
		return this.httpClient
			.post<ResponseModelBase>(`${this.path}/password/reset`, body)
			.then(this.processResponse.bind(this))
			.finally(this.finalize.bind(this))
	}

	public getUserPassport() {
		return this.httpClient
			.get<AuthenticationResponseModel>(`${this.path}/passport`)
			.then(this.processResponse.bind(this))
			.finally(this.finalize.bind(this))
	}
}

export default GateService

File: ASMR.Mobile/ASMR.Mobile/Services/GateService.cs
using System.Threading.Tasks;
using ASMR.Core.RequestModel;
using ASMR.Core.ResponseModel;
using ASMR.Mobile.Services.Abstraction;
using ASMR.Mobile.Services.Generic;

namespace ASMR.Mobile.Services
{
	public class GateService : BackEndService, IGateService
	{
		public GateService() : base("gate")
		{
		}
		
		public Task<AuthenticationResponseModel> ResendEmailAddressConfirmation(
			ResendEmailAddressConfirmationRequestModel model)
		{
			var endpoint = GetServiceEndpoint().AppendPathSegments("email-address", "resend-confirmation");
			return PostAsync<AuthenticationResponseModel, ResendEmailAddressConfirmationRequestModel>(
				endpoint, model);
		}
		
		public Task<AuthenticationResponseModel> Authenticate(SignInRequestModel model)
		{
			var endpoint = GetServiceEndpoint().AppendPathSegments("authenticate");
			return PostAsync<AuthenticationResponseModel, SignInRequestModel>(
				endpoint, model);
		}
		
		public Task<AuthenticationResponseModel> ClearSession()
		{
			var endpoint = GetServiceEndpoint().AppendPathSegments("exit");
			return PostAsync<AuthenticationResponseModel>(endpoint);
		}
		
		public Task<AuthenticationResponseModel> ForgetPassword(ForgetPasswordRequestModel model)
		{
			var endpoint = GetServiceEndpoint().AppendPathSegments("password", "forget");
			return PostAsync<AuthenticationResponseModel, ForgetPasswordRequestModel>(
				endpoint, model);
		}
		
		public Task<AuthenticationResponseModel> GetUserPassport()
		{
			var endpoint = GetServiceEndpoint().AppendPathSegments("passport");
			return GetAsync<AuthenticationResponseModel>(endpoint);
		}

		
	}
}